File: src/main/java/leetcode/p033/Solution.java
package leetcode.p033;

import java.util.Arrays;

class Solution {
    public static void main(String[] args) {
        int[] arr = {8, 9, 10, 11, 12, 13, 14, 15, 1, 2, 3, 4, 4, 4, 4, 8};
        Solution sol = new Solution();
        int idx = sol.search(arr, 15);
        System.out.println(idx);
    }

    int search(int[] nums, int target) {
        if (nums == null) return -1;
        if (nums.length == 0) return -1;
        if (nums.length == 1)
            if (nums[0] == target) return 0;
            else return -1;


        if (nums[0] < nums[nums.length-1]){ //its not rotated
            return nt1(Arrays.binarySearch(nums, 0, nums.length , target));
        }
        int idx = findbreak(nums, nums[0], 0, nums.length);
        //System.out.println(idx);
        if (target == nums[0])
            return 0;
        else if (target < nums[0]) {
            return nt1(Arrays.binarySearch(nums, idx, nums.length , target));
        } else
            return nt1(Arrays.binarySearch(nums, 0, idx , target));
    }

    int nt1 (int num) {
        if (num < 0) return -1;
        else return num;
    }


    int findbreak(int[] nums, int smallerthen, int left, int right) {
        if (left > right)
            return left;
        int mid = left + (right - left) / 2;
        int cur = nums[mid];
        if (cur > smallerthen)
            return findbreak(nums, smallerthen, mid + 1, right);
        else {
            return findbreak(nums, cur, left, mid - 1);
        }
    }


}

File: scala/src/leetcode/p033/Solution.scala
package leetcode.p033

import scala.collection.Searching.SearchResult
import scala.collection.Searching.InsertionPoint
import scala.collection.Searching.Found

object Solution {
  def search(nums: Array[Int], target: Int): Int = {
    if (nums.isEmpty) return -1
    val shift = binsearch(0, nums.size, nums, None, nums(0))
    val shiftidx = shift match {
      case Found(foundIndex)              => foundIndex
      case InsertionPoint(insertionPoint) => 0
    }
    val sarray = new ShiftedArray(nums, shiftidx)
    
    sarray.search(target, 0, nums.size)(Ordering.Int) match {
        case Found(foundIndex) => { (foundIndex + shiftidx ) % nums.size 
        } 
        case InsertionPoint(insertionPoint) => -1
    }
  }

  class ShiftedArray(nums: Array[Int], shift: Int) extends IndexedSeq[Int] {
    def apply(i: Int): Int = {
      nums((i + shift) % nums.size)
    }
    def length: Int = nums.length
  }

  def binsearch(
      l: Int,
      r: Int,
      arr: Array[Int],
      found: Option[Int],
      zeroele: Int
  ): SearchResult = {
    if (r <= l) {
      found match {
        case None        => return InsertionPoint(l)
        case Some(value) => return Found(value)
      }
    } else {
      val idx = l + (r - l - 1) / 2
      val check = arr(idx)
      if (check >= zeroele) {
        binsearch(idx + 1, r, arr, found, zeroele)
      } else {
        binsearch(l, idx, arr, Some(idx), zeroele)
      }
    }
  }
}


object test{
    @main
    def main ()={
        val arr = Array(4,5,6,7,0,1,2)
        //val arr = Array(1,3)
        println(Solution.search(arr, 0))
    }
}